dummyfs/dir.c

```diff
@@ -13,138 +13,141 @@
 
 #include <stdio.h>
 #include <stdlib.h>
-#include <sys/msg.h>
 #include <sys/stat.h>
+#include <sys/list.h>
 #include <string.h>
 #include <dirent.h>
+#include <assert.h>
 
 #include "dummyfs_internal.h"
+#include "memory.h"
 #include "dir.h"
 
 
-int dir_find(dummyfs_object_t *dir, const char *name, oid_t *res)
+static int dummyfs_dir_compare(rbnode_t *n1, rbnode_t *n2)
 {
+	dummyfs_dirent_t *i1 = lib_treeof(dummyfs_dirent_t, linkage, n1);
+	dummyfs_dirent_t *i2 = lib_treeof(dummyfs_dirent_t, linkage, n2);
 
-	dummyfs_dirent_t *e = dir->entries;
-	char *dirname, *end;
-	int len;
+	if (i1->key > i2->key) {
+		return 1;
+	}
+	else if (i1->key < i2->key) {
+		return -1;
+	}
+	else {
+		return 0;
+	}
+}
 
-	if (!S_ISDIR(dir->mode))
-		return -ENOTDIR;
 
-	if (e == NULL)
-		return -ENOENT;
+static uint32_t dummyfs_dir_hash(const char *name, const char *end)
+{
+	TRACE();
+	uint32_t key = 0;
+	size_t i = 0;
+
+	for (const char *s = name; s != end; ++s) {
+		key += (uint32_t)(*s) << (i & 0xf);
+		++i;
+	}
 
-	dirname = strdup(name);
-	end = strchr(dirname, '/');
-	if (end != NULL)
-		*end = 0;
+	return key;
+}
 
-	len = strlen(dirname);
 
-	/* Iterate over all entries to find the matching one */
-	do {
-		if (!strcmp(e->name, dirname) && !e->deleted) {
-			memcpy(res, &e->oid, sizeof(oid_t));
-			free(dirname);
-			return len;
+static dummyfs_dirent_t *dummyfs_dir_get(dummyfs_object_t *dir, const char *name)
+{
+	TRACE();
+	const char *end = strchrnul(name, '/');
+	dummyfs_dirent_t t = { .key = dummyfs_dir_hash(name, end) };
+	dummyfs_dirent_t *e = lib_treeof(dummyfs_dirent_t, linkage, lib_rbFind(&dir->dir.tree, &t.linkage));
+
+	while (e != NULL) {
+		if (strncmp(e->name, name, end - name) == 0) {
+			break;
```

Comment:
@agkaminski, look at a case like this:
```
e->name == "test"
name == "t"
end - name == 1
```
and
`strncmp()` will return the value `0`.

You should additionally check the length of the `e->name`
```suggestion
	const size_t len = strchrnul(name, '/') - name;
	dummyfs_dirent_t t = { .key = dummyfs_dir_hash(name, end) };
	dummyfs_dirent_t *e = lib_treeof(dummyfs_dirent_t, linkage, lib_rbFind(&dir->dir.tree, &t.linkage));

	while (e != NULL) {
		if ((strncmp(e->name, name, len) == 0)  && (e->name[len] == '\0')) {
			break;
```

Reply:
Thanks, fixed